### PR TITLE
Introduce the Logstash operator for ECK

### PR DESCRIPTION
**Pull request by robbavey**
This commit introduces a technical preview of the Logstash Operator for ECK

The Logstash operator introduces a Logstash CRD:

Look and feel: 

Logstash with a simple pipeline, pointing to an elasticsearch instance:

```yaml
apiVersion: logstash.k8s.elastic.co/v1alpha1
kind: Logstash
metadata:
  name: logstash-sample
spec:
  count: 1
  version: 8.7.0
  elasticsearchRefs:
    - clusterName: production
      name: elasticsearch-sample
  pipelines:
    - pipeline.id: main
      config.string: |
        input { exec { command => 'uptime' interval => 10 } } 
        output { 
          elasticsearch {
            hosts => [ "${PRODUCTION_ES_HOSTS}" ]
            ssl => true
            cacert => "${PRODUCTION_ES_SSL_CERTIFICATE_AUTHORITY}"
            user => "${PRODUCTION_ES_USER}"
            password => "${PRODUCTION_ES_PASSWORD}"
          } 
        }
```

Produces the following resources:
```sh
 ✗ kubectl get logstash logstash-sample
NAME              AVAILABLE   EXPECTED   AGE     VERSION
logstash-sample   1           1          3m34s   8.7.0
```

```sh
 kubectl tree logstash logstash-sample
NAMESPACE  NAME                                                             READY  REASON  AGE  
pr         Logstash/logstash-sample                                         -              3m57s
pr         ├─Secret/logstash-sample-logstash-es-pr-elasticsearch-sample-ca  -              3m56s
pr         ├─Secret/logstash-sample-ls-config                               -              3m57s
pr         ├─Secret/logstash-sample-ls-pipeline                             -              3m57s
pr         ├─Secret/logstash-sample-pr-elasticsearch-sample-logstash-user   -              3m57s
pr         ├─Service/logstash-sample-ls-api                                 -              3m57s
pr         │ └─EndpointSlice/logstash-sample-ls-api-ndzf5                   -              3m57s
pr         └─StatefulSet/logstash-sample-ls                                 -              3m57s
pr           ├─ControllerRevision/logstash-sample-ls-7d8687f9b              -              3m57s
pr           ├─ControllerRevision/logstash-sample-ls-c6cd667c4              -              3m56s
pr           └─Pod/logstash-sample-ls-0                                     True           3m52s

```

This operator provides support for:

* Defining `logstash.yml` in `config` or `configRef` sections of the CRD
* Integration with Elasticsearch clusters via the use of `elasticsearchRefs`, and environment variable substitution to introduce those elasticsearch references into logstash pipelines.
* Definition of `pipelines.yml` in `pipelines` or `pipelinesRef` sections of the CRD with support for pipeline definition in volume mounts
* Support for multiple pipeline and pipeline->pipeline configurations
* Support for automatic pipeline reload in logstash pods when a pipeline change is detected without triggering a full restart of the pod.
* Stack monitoring support via sending metrics and logs to a monitoring elasticsearch cluster via the use of `monitoring.logs.elasticsearchRefs` and `monitoring.metrics.elasticsearchRefs`
* Support for defining multiple services for logstash plugins. 

Logstash nodes are created as StatefulSets - we expect in later versions of the logstash operator to support persistence in Logstash nodes, including persistent queues and dead letter queues. 
 
A work in progress [PR](https://github.com/elastic/cloud-on-k8s/pull/6585) includes documentation and recipes on how to use this logstash operator. There are also samples in this PR located under `config/samples/logstash/*`

Relates: https://github.com/elastic/cloud-on-k8s/issues/1453

### CODE REVIEW

**Comment by thbkrkr** [comment]
buildkite test this -f p=gke -m t=TestSamples,t=TestVersionUpgradeOrdering,E2E_TAGS=logstash

**Comment by robbavey** [comment]
buildkite test this -f p=gke -m t=TestSamples,t=TestVersionUpgradeOrdering,E2E_TAGS=logstash

**Comment by robbavey** [comment]
@pebrc @barkbay @thbkrkr Fixed the issue with the tests with a commit this morning, and this is now ready for review. 

Thanks!

**Review by barkbay — APPROVED** [review]
Since all the commits in this PR have already been reviewed, and the new E2E tests seem to pass, I think we can merge it as is.

@robbavey please let me know if you think that's not the case and if anything else should be reviewed by the ECK team.

File: test/e2e/test/logstash/http_client.go
Comment (pebrc) [inline test/e2e/test/logstash/http_client.go:24]: I assume we capture these TODOs in issues somewhere, correct?

**Review by pebrc — APPROVED** [review]
LGTM as @barkbay said, this looks like the code we already reviewed. Let us know if there is something specific to look for. Otherwise I am really looking forward to merging this to main. 🎉

**Review by thbkrkr — APPROVED** [review]
🚀 🧪 🍀

**Comment by robbavey** [comment]
@barkbay @pebrc Nothing new to review, just wanted to get final approval from ECK team before merge

**Comment by robbavey** [comment]
@pebrc @barkbay @thbkrkr Thank you for the reviews, getting @kaisecheng and I to this point - from the Logstash side, I am happy to see this merged into `main`, but I don't have the authorization to do so... If one of you could please do the honours and press the button, please go ahead

**Comment by robbavey** [comment]
Also, are there any release notes/changelog entries that you would like me to add?

**Comment by thbkrkr** [comment]
> Also, are there any release notes/changelog entries that you would like me to add?

This is handled separately and we will ask you at that time to contribute to it, thank you.

**Pull request merged**
